source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst

```diff
@@ -0,0 +1,156 @@
+
+.. code-block:: java
+
+   public MyPoint2D withX(double newY) {
+      return new MyPoint2D(this.x, newY);
+   }
+
+This creates a new ``MyPoint2D`` object and returns it with the new y-coordinate.
+
+Creating a Data Types
+---------------------
+There are two different data types that can be made, Simple data types involve data that contains only one field (ie. a single number or string) whereas Complex data types require multiple data fields (ie. multiple strings, multiple numbers).
+
+In order to define a simple data type, the class must extend the ``SimpleDataType<DataType>`` class with the data type needed and implement the ``getDefaultValue()`` method. In this example, we will use a double as our simple data type.
+
+.. code-block:: java
+
+   public final class MyDataType extends SimpleDataType<double> {
+
+      private static final String NAME = "double";
+
+      private MyDataType() {
+         super(NAME, double.class);
+      }
+
+      @Override
+      public double getDefaultValue() {
+         return 0.0;
+      }
+
+   }
+
+The class constructor is set to private to ensure that only a single instance of the data type will exist.
+
+In order to define a complex data type, the class must extend the ``ComplexDataType`` class and override the ``fromMap()`` and ``getDefaultValue()`` methods.
+We will use our MyPoint2D class as an example to see what a complex data type class would look like.
+
+.. code-block:: java
+
+   public final class PointDataType extends ComplexDataType<MyPoint2D> {
+
+      private static final String NAME = "MyPoint2D";
+      public static final PointDataType Instance = new PointDataType();
+
+      public PointDataType() {
+         super(NAME, MyPoint2D.class);
+      }
+
+      @Override
+      public Function<Map<String, Object>, MyPoint2D> fromMap() {
+         return map -> {
+               return new MyPoint2D((double) map.getOrDefault("x", 0.0), (double) map.getOrDefault("y", 0.0));
+         };
+      }
+
+      @Override
+      public MyPoint2D getDefaultValue() {
+         // use default values of 0 for X and Y coordinate
+         return new MyPoint2D(0, 0);
+      }
+
+   }
+
+The following code above works as noted:
+
+The ``fromMap()`` method creates a new MyPoint2D using the values in the NetworkTable entry it is bound to.
+The ``getOrDefault`` method will return 0.0 if it cannot get the entry values. The ``getDefaultValue`` will return a new ``MyPoint2D`` object if no source is present.
+
+Exporting Data Type To Plugin
+-----------------------------
+In order to have the data type be recognized by the Shuffleboard, the plugin must export them by overriding the ``getDataTypes`` method.
+For example,
+
+.. code-block:: java
+
+   public class MyPlugin extends Plugin {
+
+      @Override
+      public List<DataType> getDataTypes() {
+         ImmutableList.of(PointDataType.Instance);
```

Comment:
This could also be `List.of` (added in Java 10, after Shuffleboard was first written)